include/kernels/madwf_tensor.cuh

```diff
@@ -0,0 +1,117 @@
+#pragma once
+
+#include <color_spinor_field.h>
+#include <color_spinor_field_order.h>
+#include <shared_memory_cache_helper.cuh>
+#include <reduce_helper.h>
+#include <kernels/madwf_transfer.cuh>
+#include <fast_intdiv.h>
+#include <madwf_ml.h>
+
+/**
+  @file This file contains the argument and kernel that forms a tensor outer product in the fifth
+  dimension given two input vectors,
+      T_{st} = conj(x_s * y_t),
+  where s and t are fifth dimension indices. T_{st} is a spin matrix so T has shape Ls*4-by-Ls*4.
+*/
+
+namespace quda
+{
+
+  namespace madwf_ml
+  {
+
+    template <class storage_t, int nSpin_, int nColor_>
+    struct Tensor5DReduceArg : ReduceArg<array<complex<typename mapper<storage_t>::type>, nSpin_ * nSpin_>> {
+
+      static constexpr int nSpin = nSpin_;
+      static constexpr int nColor = nColor_;
+
+      using F = typename colorspinor_mapper<storage_t, nSpin, nColor>::type;
+      using real = typename mapper<storage_t>::type;
+      using reduce_t = array<complex<real>, nSpin * nSpin>;
+      using Vector = ColorSpinor<real, nColor, nSpin>;
+
+      const F x; // xput vector field
+      const F y; // yput vector field
+
+      const int_fastdiv Ls_x; // length of 5th dimension of the x field
+      const int_fastdiv Ls_y; // length of 5th dimension of the in field
+
+      const int volume_4d_cb;
+
+      const int nParity;
+
+      static constexpr unsigned int max_n_batch_block = 1u;
+
+      Tensor5DReduceArg(const ColorSpinorField &x, const ColorSpinorField &y) :
+        ReduceArg<reduce_t>(dim3(x.VolumeCB() / x.X(4), x.SiteSubset(), x.X(4) * y.X(4)), x.X(4) * y.X(4)),
+        x(x),
+        y(y),
+        Ls_x(x.X(4)),
+        Ls_y(y.X(4)),
+        volume_4d_cb(y.VolumeCB() / y.X(4)),
+        nParity(y.SiteSubset())
+      {
+        if (volume_4d_cb != static_cast<int>(x.VolumeCB() / x.X(4))) {
```

Comment:
Do we need this check for both `x` and `y`?

Reply:
In the initialization list `volume_4d_cb(y.VolumeCB() / y.X(4))` so we only need to check for `x`.